dftimewolf/lib/collectors/gsheets.py

```diff
@@ -0,0 +1,254 @@
+               name: Optional[str] = None,
+               critical: bool = False) -> None:
+    """Initializes a Google Sheets collector."""
+    super(GoogleSheetsCollector, self).__init__(
+        state, name=name, critical=critical)
+    self._credentials = None
+    self._spreadsheet_id = ''
+    self._sheet_names: List[str] = []
+    # These are mandatory columns required by Timesketch.
+    self._mandatory_columns = ['message', 'datetime', 'timestamp_desc']
+    self._all_sheets = True
+    self._validate_columns = True
+
+  # pylint: disable=arguments-differ
+  def SetUp(self,
+            spreadsheet: str,
+            sheet_names: List[str],
+            validate_columns: bool = True) -> None:
+    """Sets up a a Google Sheets collector.
+
+    Args:
+      spreadsheet: ID or URL of the sheet to pull data from
+      sheet_names: List of sheet names inside the spreadsheet to parse. 'All'
+        will parse all sheets inside a spreadsheet.
+      validate_columns: Check if mandatory columns required by Timesketch is
+        present in the sheets.
+    """
+    self._credentials = self._GetCredentials()
+    self._spreadsheet_id = self._ValidateSpreadSheetId(spreadsheet)
+    self._sheet_names = sheet_names
+    if not sheet_names:
+      self._all_sheets = True
+    self._validate_columns = validate_columns
+
+  def Process(self) -> None:
+    """Copies entries from Google Sheets."""
+
+    try:
+      # Retrive list of sheets in the spreadsheet
+      service = self._BuildSheetsResource(self._credentials)
+      # Pylint can't see the spreadsheets method.
+      # pylint: disable=no-member
+      result = service.spreadsheets().get(
+          spreadsheetId=self._spreadsheet_id).execute()
+      spreadsheet_title = result.get('properties', {}).get('title')
+      sheets = (result.get('sheets', []))
+
+      for sheet in sheets:
+        if not sheet.get('properties'):
+          continue
+
+        sheet_title = sheet.get('properties', {}).get('title')
+
+        if not self._all_sheets and sheet_title not in self._sheet_names:
+          continue
+
+        self.logger.info('Parsing sheet: {0:s}'.format(sheet_title))
+
+        df = self._ExtractEntiresFromSheet(self._spreadsheet_id, sheet_title)
+
+        if df is None or df.empty:
+          continue
+
+        output_file = tempfile.NamedTemporaryFile(
+            mode='w', delete=False, encoding='utf-8', suffix='.csv')
+        output_path = output_file.name
+        self.logger.info(
+            'Downloading results of sheet "{0:s}" to {1:s}'.format(
+                sheet_title, output_path))
+
+        df.to_csv(index=False, na_rep='NaN', path_or_buf=output_file)
+
+        self.logger.success(
+            'Downloaded results of sheet "{0:s}" to {1:s}'.format(
+                sheet_title, output_path))
+        output_file.close()
+        sheet_csv_file = containers.File(
+            name='{0:s}_{1:s}'.format(spreadsheet_title, sheet_title),
```

Comment:
Could we set the name attribute as the spreadsheet ID and the description as {spreadsheet_title}_{sheet_title} instead? 

That way we can do ```state.GetContainers(containes.File)``` and filter by spreadsheet ID in a downstream module (e.g. collector) later in the recipe without having to resolve the spreadsheet/sheet title values again.

Reply:
Yup, will do that